---
title: "Using Seurat's manifold aligment with SWNE"
author: "Yan Wu"
date: "5/21/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This is a walkthrough demonstrating how to generate SWNE plots alongside the [Seurat](http://satijalab.org/seurat/) [manifold alignment pipeline]() from four pancreas [datasets](http://bit.ly/IAexpmat) generated using different single cell RNA-seq technologies.

To save time we will be using the pre-computed Seurat object `pancreas_integrated_seurat.Robj`, which can be downloaded [here](ftp://genome-miner.ucsd.edu/swne_files/pancreas_integrated_seurat.Robj).


First let's load the required libraries
```{r message=FALSE,warning=FALSE}
library(Seurat)
library(swne)
```


Let's load the Seurat object
```{r}
se.obj <- readRDS("pancreas_integrated_seurat.Robj")
```


We can "reconstruct" the gene expression matrix from the aligned correlated components and the corresponding gene loadings. This is not equivalent to the original gene expression matrix, as it only contains the reduced information from the correlated components, but it is batch-corrected and something we can run NMF on.
```{r}
cca.aligned.expr <- ExtractDebatchedSeurat(se.obj)
```


Once we have the "reconstructed matrix", we can run NMF and embed the components
```{r message=FALSE,warning=FALSE}
k <- 14
loss <- "mse"
n.cores <- 16

nmf.res <- RunNMF(cca.aligned.expr, k = k, alpha = 0, init = "ica", n.cores = n.cores,
                  loss = loss, init.zeros = "random")

swne.embedding <- EmbedSWNE(nmf.res$H, se.obj@snn, alpha.exp = 2, snn.exp = 0.1, 
                            n_pull = 4, dist.use = "IC", min.snn = 0.0)
```


We project the full gene expression matrix to get gene loadings for the full set of genes
```{r message=FALSE,warning=FALSE}
norm.counts <- ScaleCounts(se.obj@raw.data[,se.obj@cell.names], method = "log")
nmf.res$W <- ProjectFeatures(norm.counts, nmf.res$H, loss = loss, n.cores = n.cores)
```


We use these gene loadings to identify key genes to embed, and we hide all the factors
```{r}
gene.factor.df <- SummarizeAssocFeatures(nmf.res$W, features.return = 1)
genes.embed <- unique(gene.factor.df$feature)

swne.embedding <- EmbedFeatures(swne.embedding, nmf.res$W, genes.embed, n_pull = 4)
swne.embedding$H.coords$name <- ""
```


We pull out the clusters and batches
```{r}
clusters <- se.obj@ident; names(clusters) <- se.obj@cell.names;
batch <- factor(se.obj@meta.data$tech); names(batch) <- se.obj@cell.names
```


We can then create the SWNE plot
```{r}
PlotSWNE(swne.embedding, alpha.plot = 0.4, sample.groups = clusters, do.label = T, 
         label.size = 3.5, pt.size = 1.25, show.legend = F, seed = 3252354)
```


We also can show that there are no batch effects
```{r}
PlotSWNE(swne.embedding, alpha.plot = 0.4, sample.groups = batch, do.label = F, 
         label.size = 3.5, pt.size = 1.25, show.legend = T, seed = 3252354)
```


t-SNE plot for comparison
```{r}
tsne.emb <- GetCellEmbeddings(se.obj, "tsne")
PlotDims(tsne.emb, sample.groups = clusters, show.legend = F, seed = 3252354)
```